design/vela-core/rollout-design.md

```diff
@@ -0,0 +1,269 @@
+	// +optional
+	NumBatches *int32 `json:"numBatches,omitempty"`
+
+	// The exact distribution among batches.
+	// mutually exclusive to NumBatches
+	// +optional
+	RolloutBatches []RolloutBatch `json:"rolloutBatches,omitempty"`
+
+	// All pods in the batches up to the batchPartition (included) will have
+	// the target resource specification while the rest still have the source resource
+	// This is designed for the operators to manually rollout
+	// Default is the the number of batches which will rollout all the batches
+	// +optional
+	BatchPartition *int32 `json:"lastBatchToRollout,omitempty"`
+
+	// RevertOnDelete revert the rollout when the rollout CR is deleted, default is false 
+	//+optional 
+	RevertOnDelete bool `json:"revertOnDelete,omitempty"`
+
+	// Paused the rollout, default is false 
+	//+optional 
+	Paused bool `json:"paused,omitempty"`
+
+	// RolloutWebhooks provides a way for the rollout to interact with an external process
+	// +optional
+	RolloutWebhooks []RolloutWebhook `json:"rolloutWebhooks,omitempty"`
+
+	// CanaryMetric provides a way for the rollout process to automatically check certain metrics
+	// before complete the process
+	// +optional
+	CanaryMetric []CanaryMetric `json:"canaryMetric,omitempty"`
+}
+```
+
+## User Experience Workflow
+OAM rollout experience is different from flagger in a few different ways and here are the
+ implications on its impact on the user experience.
+- We assume that the resources it refers to are **_immutable_**. In contrast, flagger watches over a
+  target resource and reacts whenever the target's specification changes.
+    - The trait version of the controller refers to componentRevision.
+    - The application version of the controller refers to immutable application. 
+- The rollout logic **_works only once_** and stops after it reaches a terminal state. One can
+  still change the rollout plan in the middle of the rollout as long as it does not change the
+  pods that are already updated. 
+- The applicationDeployment controller only works with applications with 1 component for now.
+- Users should rely on the rollout CR to do the actual rollout which means they
+ shall set the `replicas` or `partition` field of the new resources to the starting value
+ indicated in the [detailed rollout plan design](#rollout-plan-work-with-different-type-of-workloads).
+- 
+    
+ 
+## Notable implementation level details
+Here are some high level implementation decisions
+
+### Rollout on different resources
+As we mentioned in the introduction section, we will implement two rollout controllers that work
+on different levels. At the end, they both emit an in-memory rollout plan object which includes
+references to the target and source kubernetes resources that the rollout planner will execute
+upon. For example, the applicationDeployment controller will get the component from the
+ application and extract the real workload from it before passing it to the rollout plan object.
```

Comment:
So the trait controller will only refer to the upper workload and vela-core/OAM runtime will do the extraction of the real underlying workload?

Reply:
the controller will do the extraction, not the OAM runtime